docs/madr/1-reassign-evolve.md

```diff
@@ -0,0 +1,192 @@
+    participant AntRegistry
+    participant ARIOProcess
+
+
+    Owner ->> ARIOGateway: Get Module ID for process from GQL
+    ARIOGateway -->> Owner: Module ID Result
+
+
+    alt No upgrade needed Module ID from process meta is up to date
+
+        else
+
+        Owner->>OldAnt: Get state
+        OldAnt-->>Owner: Return state
+
+        Owner->>NewAnt: Spawn new ant with old state
+        NewAnt->>AntRegistry: Register ANT notice
+        Owner->>NewAnt: Get State to verify responsiveness and Module ID to verify correct deployment
+
+        Owner->>OldAnt: Sign and accept Reassign message
+        OldAnt->>ARIOProcess: Forward Reassign Message
+        ARIOProcess-->>OldAnt: Reassign-Notice
+
+        Owner->>ARIOProcess: Check for Reassignment success and notify user
+
+
+    end
+```
+
+### Positive Consequences
+
+- **Performance Improvement:** Reducing dry runs significantly speeds up ANT
+  loading.
+- **Immutable Module Versioning:** Eliminates issues with heuristic version
+  analysis.
+  - Downstream clients can identify and validate the capabilities of a module
+    (and by extension the ANT itself) has by maintaining a map of modules that
+    are capables of executing target workflows
+  - Gives us the ability to guarantee that we will not brick the ANT, because we
+    can spawn it then validate it, unlike with Eval where we cannot eval first,
+    then test after.
+- **Reduced CU Load:** Fewer computations per ANT.
+- **Gateway ANT Module Whitelisting:** Gateways can decide which ANT Modules
+  they support.
+
+### Negative Consequences
+
+- **Complicates Memory Usage on ANT Registry:** More ANTs means more state in
+  the ANT registry, which currently doesn't have a cleanup process.
+- **Potential Integration Issues:** Existing integrations relying on fixed ANT
+  IDs may break.
+- **State Limitations:** Large ANT states (e.g., >2000 undernames) may fail to
+  bootstrap on spawn, meaning we need to understand the limitations there and
+  decide what amounts we wish to support.
+- **MU cranking risks**: Its possible a module fails to instantiate/boot
+  properly - the defense there is to query for its state to ensure
+  responsiveness once spawned. Users will not be presented to reassign a name if
+  spawning fails, or a newly created process is not as responding as expected.
+- **Custom ANTs will be lost**: If a user has a custom ANT implementation
+  (meaning code that is outside of our control is loaded in) then this code and
+  state will not be carried over.
+
+## Pros and Cons of the Options
+
+### Current Approach
+
+- `+` Already implemented and functional.
+- `+` Ensures workflow compatibility via handler checks and dryruns.
+- `-` Multiple dry runs (in our case 3) increases computational cost for CUs.
+- `-` Upgrade process relies on heuristic version checks.
+- `-` No way to assert capabilities and schema compliance of the ANT without
+  multiple dryruns.
+- `-` Relying on Eval poses a security risk.
+
```

Comment:
`- Evals don't allow to validate behavior of the ANT after eval occurs`

Reply:
https://github.com/ar-io/ar-io-ant-process/pull/74/commits/4c64200eda9969fab608f83f83370f8088d93d64